### PR TITLE
v2 volume: support arm64 platform

### PR DESCRIPTION
**Pull request by derekbit**
Longhorn/longhorn#6021

### CODE REVIEW

**Comment by AaronDewes** [comment]
Is there a plan when this will be merged? I'd love to try Longhorn's v2 data engine on arm64.

File: package/Dockerfile [hunk 1]
```diff
@@ -66,32 +75,36 @@
...
+RUN zypper -n install nfs-client nfs4-acl-tools cifs-utils sg3_utils \
+    iputils iproute2 qemu-tools e2fsprogs util-linux-systemd python3-pyelftools libcmocka-devel&& \
+    rm -rf /var/cache/zypp/*
 
 # Install SPDK dependencies
 COPY --from=builder /usr/src/spdk/scripts /spdk/
-RUN if [ "${ARCH}" == "amd64" ]; then \
-        sh /spdk/pkgdep.sh; \
-        rm -rf /tmp/spdk; \
+RUN if [ "${ARCH}" = "amd64" ] || [ "${ARCH}" = "arm64" ]; then \
```
Comment (innobead) [inline package/Dockerfile:100]: If both are supported, let's just remove this if statement.

**Review by innobead — CHANGES_REQUESTED** [review]
LGTM. Just one comment.

File: package/Dockerfile (hunk 1 above)
Comment (derekbit) [reply package/Dockerfile:100]: We won't support s390x anymore, right?

**Comment by derekbit** [comment]
v2 volume is successfully attached and running on arm64 platfrom.
```
NAME                                       STATE      ROBUSTNESS   SCHEDULED   SIZE         NODE                AGE
pvc-425b0e1b-e76d-4014-af82-bfa1e74895b7   attached   healthy                  2147483648   c3-large-arm64-01   38s
```

File: package/Dockerfile (hunk 1 above)
Comment (innobead) [reply package/Dockerfile]: Yeah. We are going to remove s390.

**Comment by derekbit** [comment]
e2e test
https://ci.longhorn.io/job/private/job/longhorn-tests-regression/5540/

File: package/Dockerfile [hunk 2]
```diff
@@ -54,6 +54,15 @@ RUN git clone https://github.com/longhorn/spdk.git ${SPDK_DIR} --recursive && \
...
         make install; \
+    else if [ ${ARCH} == "arm64" ]; then \
+        cd ${SPDK_DIR} && \
+        git checkout ${SPDK_COMMIT_ID} && \
+        git submodule update --init && \
+        ./scripts/pkgdep.sh && \
+        ./configure --target-arch=native --disable-tests --disable-unit-tests --disable-examples && \
+        DPDKBUILD_FLAGS="-Dplatform=generic" make -j$(nproc) && \
+        make install; \
+    fi; \
```
Comment (innobead) [inline package/Dockerfile:65]: redundant `fi`

**Review by innobead — CHANGES_REQUESTED** [review]

File: package/Dockerfile [hunk 3]
```diff
@@ -64,7 +85,7 @@ RUN git clone https://github.com/linux-nvme/nvme-cli.git ${NVME_CLI_DIR} && \
     git checkout ${NVME_CLI_COMMIT_ID} && \
     meson setup --force-fallback-for=libnvme .build && \
     meson compile -C .build && \
-    meson install -C .build --destdir /usr/local/nvme
+    meson install -C .build
```
Comment (c3y1huang) [inline package/Dockerfile:88]: Redundant trailing spaces.

File: package/Dockerfile [hunk 4]
```diff
@@ -82,31 +103,34 @@ RUN zypper -n addrepo --refresh https://download.opensuse.org/repositories/syste
...
-        rm -rf /spdk; \
-    fi
+RUN sh /spdk/pkgdep.sh && rm -rf /spdk
 
 # Copy pre-built binaries from builder
 COPY --from=builder \
     /usr/local/bin/grpc_health_probe \
     /usr/local/bin/spdk_* \
+    /usr/local/bin/go-spdk-helper \
+    /usr/local/bin/
```
Comment (c3y1huang) [inline package/Dockerfile:118]: Doesn't `COPY /usr/local/bin/` imply copying all files within the directory, including the `grpc_health_probe`, `spdk_*` and `go-spdk-helper`? Same question for `/usr/local/sbin/` and `/usr/sbin/`.

Comment (derekbit) [reply package/Dockerfile:118]: The last line `/usr/local/bin/` is the destination directory.

Comment (c3y1huang) [reply package/Dockerfile:118]: I see, it's clear.

**Review by innobead — APPROVED** [review]

File: package/Dockerfile (hunk 2 above)
Comment (derekbit) [reply package/Dockerfile]: Updated. Thank you.

File: package/Dockerfile (hunk 3 above)
Comment (derekbit) [reply package/Dockerfile]: Updated. Thank you.

**Review by innobead — APPROVED** [review]

**Review by innobead — APPROVED** [review]

**Comment by innobead** [comment]
@derekbit well done.

**Pull request merged**